### PR TITLE
Introduce krnl.load and krnl.store

### PR DESCRIPTION
**Pull request by tungld**
This patch introduces krnl.load and krnl.store  ops. These ops will replace std.load/store or affine.load/store when we lower ONNX ops to krnl dialect. 

With krnl.load/store, we don't need to care about whether access indices are affine or not. The lowering of krnl.load/store to affine and std dialects will automatically check access indices and decide which one to use.

Also, we can remove the dependency of the krnl.iterate op to affine (say, no need of AffineScope trait in the krnl.iterate definition). affine.load/store will be emitted at the same time affine.for is emitted, which avoids potential errors related to affine related optimization passes.

### CODE REVIEW

File: src/Transform/BundleMemoryPools.cpp
Comment (doru1004) [inline src/Transform/BundleMemoryPools.cpp:323]: When is the lowering of krnl.load/store performed?

Comment (tungld) [reply src/Transform/BundleMemoryPools.cpp:323]: When we lower krnl to affine, i.e. inside convert-krnl-to-affine pass.

**Comment by AlexandreEichenberger** [comment]
While technically changing the Load/store to krnl operations is feasible, I believe we need a better understanding of why we do it, what we gain, and what we may lose. In particular, having custom operation may prevent optimizations that are written in MLIR as it does not know about the krnl dialect. I believe you had an example of a strange behavior witnessed with the affine/std memory ops, and I think we need to collectively think about that example, and the pluses/minues to go to that approach.

On an unrelated ops, I tried to systematically use krnl.dim and it lost a lot of std.dim optimizations, which are crucial to bundling. I am not saying the same thing will happen here, but just want to raise the awareness of the cost of going custom, when there are alternatives. 

We can start the discussion here, and maybe have an in-depth presentations at some time so that we are all onboard with this PR. Thanks

**Comment by doru1004** [comment]
@AlexandreEichenberger I agree.

I think one thing we need to test is if we use krnl.load/krnl.store and then we end up lowering them to std.load/std.store and canonicalize then we end up with the exact same code as if we just used std.load/std.store directly.

This observation is also valid for krnl.dim. We can "abuse" krnl.dim as long as we lower it fairly early and give the std optimization a chance to run on the std.dims we obtain from the lowered krnl.dims.

**Comment by tungld** [comment]
As you are aware of, we had to disable affine-fusion in [MainUtils.cpp](https://github.com/onnx/onnx-mlir/blob/master/src/MainUtils.cpp#L460), it is because the program we emit is not compatible with MLIR affine related optimizations such as affine-loop-fuse, affine-loop-tile, etc. So we cannot do any MLIR affine related optimizations. To make it compatible, I had to disable all canonicalizations that happen before convert-krnl-to-affine, which is a very bad thing.

The same thing happens when we run our LSTM model where we need to call `memref-normalize`, we also need to disable all canonicalizations before convert-krnl-to-affine to compile the model.

With the above issues, I think it is not good to have affine.load/affine.store when affine.for has not appeared. So, by introducing krnl.load/krnl.store, we lower them to affine.load/store at the same time we lower krnl.iterate to afffine.for, which, I believe, makes MLIR optimization valid. With krnl.load/krnl.store, I can call arbitrary MLIR affine related optimizations without any problem.

**Comment by doru1004** [comment]
@tungld just to make sure, if you emit affine load/stores, canonicalize and then emit affine.for loops later and only after that you attempt to perform affine-loop-fuse and affine-loop-tile affine optimizations, they fail? Does this failure happen in all cases or just some special cases?

It would be good to know what it is that canonicalization removes that those optimizations need so we can avoid such situations in the future.

**Comment by tungld** [comment]
>  if you emit affine load/stores, canonicalize and then emit affine.for loops later and only after that you attempt to perform affine-loop-fuse and affine-loop-tile affine optimizations, they fail?
Yes. It happens all the time. The root cause is affine.load/store's canonicalization. The ops are inside krnl.iterate's body which is marked with AffineScope, so canonicalization will promote all access indices of affine.load/store to symbol. So far, in MLIR, marking a custom op with AffineScope is the only way to allow affine.load/store uses variables as access indices.

**Comment by doru1004** [comment]
> so canonicalization will promote all access indices of affine.load/store to symbol

Sounds like this might be the actual problem.

File: src/Dialect/ONNX/IndexExpr.hpp
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/IndexExpr.hpp:289]: Is there still a need for the regular "createLoadOp" and "createStoreOp"? I would think that they can be removed

File: src/Conversion/KrnlToAffine/KrnlToAffine.cpp
Comment (AlexandreEichenberger) [inline src/Conversion/KrnlToAffine/KrnlToAffine.cpp:221]: Saw that you checked that the right affine/std op is generated, thanks!

File: src/Dialect/ONNX/IndexExpr.cpp
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/IndexExpr.cpp:317]: I have seen other occurrence of create<AffineLoadOp> in IndexExpr.cpp, can you double check that they are all converted to the krnl versions?

File: test/mlir/krnl/krnl_bundle_lstm_memory_pool.mlir
Comment (AlexandreEichenberger) [inline test/mlir/krnl/krnl_bundle_lstm_memory_pool.mlir:33]: Could you take this test, and feed it to krnl to affine, and make sure all of the krnl stores are transformed to the affine as seen on the original test? Visual inspection would be enough... if you wanted to make it an official test, that would be fine too.

**Review by AlexandreEichenberger — COMMENTED** [review]
@tungld I now better understand the approach, and it does indeed feel like it's the right thing to do. That was quite a change. From checking on my own upstream/master, there seem to be some create<AffineLoadOp> that are in IndexExpr that did not show up as modified to KrnlLoadOp; do you mind checking that you covered all of them?

Also, is it true that you now don't anticipate to see any std or affine load during krnl gen? If so, should we mark these operations are forbidden? Just for sanity? Because someone in the future may erroneously gen the old load/store?

**Comment by doru1004** [comment]
When you emit krnl.load/store do you emit it just in place of the affine load/stores?

Do we have operations where we emit both affine and standard load stores or do we emit only affine load/stores?

It would be good to emit standard loads and store as we do now and only provide krnl load/store for affine load/stores. Benefits:
- the lowering of the krnl load/store would become very easy, just transform to affine load/store
- standard load/stores can still be optimized

File: test/mlir/krnl/krnl_bundle_lstm_memory_pool.mlir
Comment (tungld) [reply test/mlir/krnl/krnl_bundle_lstm_memory_pool.mlir:33]: @AlexandreEichenberger I fed it krnl-to-affine and saw all krnl.load/store were transformed to affine.load/store as seen on the original test. Memory pool bundling is done at the krnl level, so it is good to see that krnl.load/store does not affect the correctness of the bundling here, instead of lower to affine.

File: src/Dialect/ONNX/IndexExpr.hpp
Comment (tungld) [reply src/Dialect/ONNX/IndexExpr.hpp:282]: I removed them. Thanks!

File: src/Dialect/ONNX/IndexExpr.cpp
Comment (tungld) [reply src/Dialect/ONNX/IndexExpr.cpp:283]: Oops, thanks for pointing it out! Fixed them.

**Comment by tungld** [comment]
@doru1004, krnl.load/store will replace both std.load/store and affine.load/store. The lowering of krnl.load/store is able to detect affine-compatiable access indices and decide to lower to std or affine. The detection is very easy by just calling a built-in MLIR method.

> It would be good to emit standard loads and store as we do now and only provide krnl load/store for affine load/stores.

I am afraid that, in this situation, users who emits code for onnx-to-krnl need to care about the difference between using std and krnl load/store.

**Comment by tungld** [comment]
> Also, is it true that you now don't anticipate to see any std or affine load during krnl gen? If so, should we mark these operations are forbidden? Just for sanity? Because someone in the future may erroneously gen the old load/store?

Yes, marked them illegal when generating krnl.

**Comment by doru1004** [comment]
> The detection is very easy by just calling a built-in MLIR method.

Is there any difference between the cases we would emit by hand and the cases the MLIR method would successfully identify i.e. are there situation which the method would not be able to identify as being affine?

File: test/mlir/krnl/krnl_bundle_lstm_memory_pool.mlir
Comment (AlexandreEichenberger) [reply test/mlir/krnl/krnl_bundle_lstm_memory_pool.mlir:33]: great, thanks for visually inspecting the code.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, @doru1004 do you want to approve as well?

**Comment by doru1004** [comment]
@AlexandreEichenberger waiting for one more answer on a question I ask earlier.

**Comment by AlexandreEichenberger** [comment]
@tungld , FYI, Doru is waiting for one more answer. Got the answers for mine, so once Doru is on board, we should be good to go.

**Comment by tungld** [comment]
> Is there any difference between the cases we would emit by hand and the cases the MLIR method would successfully identify as affine i.e. are there situations which the method would not be able to identify as being affine?

@doru1004 As I saw when changing lit tests in this patch, I saw no difference so far, the method correctly produced std or affine as we did by hand. The method has been using in MLIR affine-related optimization passes to do affine transformation, so I expect that it is able to identity affine well.

**Comment by doru1004** [comment]
> > Is there any difference between the cases we would emit by hand and the cases the MLIR method would successfully identify as affine i.e. are there situations which the method would not be able to identify as being affine?
> 
> @doru1004 As I saw when changing lit tests in this patch, I saw no difference so far, the method correctly produced std or affine as we did by hand. The method has been using in MLIR affine-related optimization passes to do affine transformation, so I expect that it is able to identity affine well.

Sounds great! Thanks for the clarification.

**Review by doru1004 — APPROVED** [review]
LGTM

**Comment by AlexandreEichenberger** [comment]
@tungld sorry #484 went before you, and I believe it must be introducing a Load/Store that result in complains. Will make our best to not have PR go before yours again.

**Comment by doru1004** [comment]
@AlexandreEichenberger just a gentle reminder to revert the title change of the patch before you merge. :)

**Comment by tungld** [comment]
@AlexandreEichenberger No problem, we could see failures when some lowering tries to use std and affine load/store.

**Comment by AlexandreEichenberger** [comment]
Thanks for your patience. You may now resume merging to master.

**Pull request merged**
